packages/core/notifications/core-notifications-browser-internal/src/toasts/deduplicate_toasts.tsx

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Toast } from '@kbn/core-notifications-browser';
+import { EuiNotificationBadge } from '@elastic/eui';
+import React from 'react';
+
+export interface DeduplicateResult {
+  toasts: Toast[];
+  idToToasts: Record<string, Toast[]>;
+}
+
+/**
+ * Collects toast messages to groups based on the `getKeyOf` function,
+ * then represents every group of message with a single toast
+ * @param allToasts
+ * @return the deduplicated list of toasts, and a lookup to find toasts represented by their first toast's ID
+ */
+export function deduplicateToasts(allToasts: Toast[]): DeduplicateResult {
+  const toastGroups = groupByKey(allToasts);
+
+  const distinctToasts: Toast[] = [];
+  const idToToasts: Record<string, Toast[]> = {};
+  for (const toastGroup of Object.values(toastGroups)) {
+    idToToasts[toastGroup[0].id] = toastGroup;
+    if (toastGroup.length === 1) {
+      distinctToasts.push(toastGroup[0]);
+    } else {
+      distinctToasts.push(mergeToasts(toastGroup));
+    }
+  }
+
+  return { toasts: distinctToasts, idToToasts };
+}
+
+/**
+ * Extracts a key from a toast message
+ * these keys decide what makes between different toasts, and which ones should be merged
+ * Ideally we want different (header+text) combinations to appear as separate toast messages
+ * @param toast
+ */
+function getKeyOf(toast: Toast): string {
+  if (isString(toast.title) && isString(toast.text)) {
+    return toast.title + ' ' + toast.text;
+  } else if (isString(toast.text)) {
+    return toast.text;
+  } else if (isString(toast.title)) {
+    return toast.title;
+  } else {
+    // Both title & text is missing or a mount function
+    return 'KEY_' + toast.id.toString();
+  }
+}
```

Comment:
Given how the current API works this approach makes sense - it's a very welcome fix!

However, I am concerned that we are fixing this problem at the wrong level: with the current approach toasts with very similar, but subtly different messages, could still cause an explosion of toast messages bc we are lacking knowledge of intention at this point (when is something "the same"?)

We can draw from React's approach to deduplicating elements in an array:

```
<>
 <El key={id1} />
 <El key={id2} />
 <El key={id2} /> 
</>
```

Consumers could optionally provide a unique group key for toasts (only for toasts we know could "spam") that we can use to correctly stack them. This could be a new option in the `ToastInput` interface.

The other part of the work, then, is to actually have UX/UI to un-stack or expand the stack (that needs to happen in EUI).

If we wanted to build this using existing components maybe we can consider:

1. Asking consumers for a group key
2. Internally stacking based on group key
3. Show one notification with an `<EuiButtomEmpty>See X more</EuiButtonEmpty>` that opens an `<EuiFlyout>` or `<EuiModal>` that lists all of the notifications...

---

All that being said, I'm OK with a "good enough" approach bc I think this toast storm problem is really annoying. Just wanted to get some more input and alignment before approving.

Reply:
We ultimately went for the cheap, easy solution (without needing a change from all consumer sites) to band-aid what we have now. As I could see from back 2020, this idea (among others) was raised, but ultimately no implementation, I assume because of the barrier to implementing a significant change. Let's start small, and introduce more sophisticated grouping later on.